### PR TITLE
[ROCm][Perf][DeepSeek V4] Enable heterogeneous MXFP4/FP8 fused shared expert

### PR DESCRIPTION
**Pull request by Fangzhou-Ai**
## Summary

Enable the native DeepSeek-V4 shared expert to use AITER's heterogeneous fused-shared-expert kernel on ROCm gfx950.

The checkpoint stores 384 routed experts as MXFP4 but stores the shared expert as serialized FP8 E4M3 with 128x128 E8M0 scales. This change keeps both formats intact:

- retain the native shared-expert FP8 weights instead of converting them to MXFP4;
- losslessly expand the 128x128 E8M0 scale grid to the kernel's per-32 layout and pad the intermediate dimension with neutral zero blocks;
- append semantic expert 384 to the top-k metadata while passing its FP8 weights and scales separately from the MXFP4 routed tensors;
- thread the heterogeneous contract through the existing AITER custom-op and fused-MoE adapters without changing the existing float custom-op ABI; and
- fall back to the separate shared MLP when the installed AITER does not expose the heterogeneous interface or any validated configuration guard is not met.

The current PR head is `c1005d4d4` on current upstream `main`. The series changes five paths: three implementation files and two focused test files.

## Why this replaces the previous implementation

The previous version of #48728 dequantized every native FP8 shared-expert matrix and requantized it to MXFP4 before appending it. Real-weight measurements showed about 12% relative-L2 reconstruction error, and the full 20-shot GSM8K run exposed a generation-quality/extraction regression. Matching AITER's MXFP4 quantizer did not recover the accuracy.

The heterogeneous kernel in ROCm/aiter#4269 removes that conversion. Routed experts remain MXFP4 and the shared expert remains FP8 within one fused stage-1/stage-2 schedule, preserving native precision without giving up the FSE performance path.

## Enablement and fallback

The optimization requires all of the following:

- ROCm gfx950;
- an AITER version containing ROCm/aiter#4269;
- AITER MoE plus `VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS=1`;
- TP=8 with no EP, EPLB, or weight offload;
- DeepSeek-V4's validated 384-routed-expert, top-k=6, one-shared-expert geometry;
- MXFP4 routed experts and serialized 128x128 block-FP8/E8M0 shared weights; and
- BF16 SiLU model execution through the AITER MoE backend.

If any guard fails, vLLM logs the reason at debug level and keeps the existing separate shared-expert MLP. A signature capability probe also prevents an older AITER installation from receiving the new arguments.

This PR remains Draft until the AITER dependency is accepted and available.

## Existing work and duplicate check

Required open-PR searches were run for `DeepSeek V4 heterogeneous shared expert AITER`, `FP8 MXFP4 fused shared expert`, and `DeepSeek V4 FSE ROCm`. No separate heterogeneous-FSE enablement PR exists. The only direct duplicate was this already-open #48728, so it was updated in place rather than opening a second PR.

vllm-project/vllm#48044 overlaps in DeepSeek-V4 shared-expert model wiring but targets AMD-Quark checkpoints whose shared expert is already MXFP4. This PR targets the native checkpoint whose shared expert is FP8 and now requires format-heterogeneous execution. The two changes must be reconciled if #48044 merges first; #48728 remains a coordinated follow-up rather than an independent competing implementation. No issue number was supplied, so the issue-specific duplicate lookup was not applicable.

## Accuracy

Post-fix validation at `c1005d4d4` used DeepSeek-V4-Pro TP8 with AITER
`0.1.20.dev13+gd5821d95e.d20260730`, all **1,319 GSM8K samples**,
20-shot prompts, concurrency 64, and temperature 0:

| Metric | Result |
|---|---:|
| Strict exact match | 1273/1319 (`0.965125 ± 0.005053`) |
| Flexible extract | 1272/1319 (`0.964367 ± 0.005106`) |

The strict score passes the 94% acceptance threshold and is two correct samples
above the directly comparable prior 20-shot run. All 1,319 API requests returned
HTTP 200, with no E8M0-scale, worker, or ROCm errors.

## 8K-input / 1K-output serving performance

Post-fix validation at `c1005d4d4` used DeepSeek-V4-Pro on 8x gfx950,
TP=8, random inputs at the harness's 0.8 input-length ratio, 1K requested output,
10 measured prompts per concurrency, and 2 warmup prompts per concurrency:

| Concurrency | Completed | Output tok/s | Total tok/s | Mean TTFT (ms) | Mean TPOT (ms) | Duration (s) |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 10/10 | 62.08 | 555.69 | 416.05 | 15.68 | 149.86 |
| 4 | 40/40 | 206.01 | 1851.93 | 490.66 | 18.43 | 178.65 |
| 8 | 80/80 | 373.83 | 3325.38 | 571.50 | 20.37 | 198.14 |
| 16 | 160/160 | 591.44 | 5337.61 | 745.25 | 25.37 | 247.58 |
| 32 | 320/320 | 867.49 | 7757.40 | 1263.79 | 34.57 | 341.38 |
| 64 | 640/640 | 1095.70 | 9876.42 | 1841.94 | 55.48 | 538.40 |

At concurrency 4, output throughput was 16.9% above the strongest directly
comparable recent local result (176.23 tok/s), while mean TPOT was 14.7% below
the best comparable result (21.599 ms).

Because AITER was intentionally installed with `PREBUILD_KERNELS=0`, the first
concurrency-32 attempt triggered a one-time CKTile build during measurement and
exceeded vLLM's 300-second model-RPC timeout. That cold attempt was discarded
(145/320 completed). After the JIT artifact finished building outside the server,
concurrency 32 and 64 were rerun from a clean server and completed 320/320 and
640/640. The final server log contained 1,153 HTTP 200 responses and no scale,
worker, ROCm, timeout, or HTTP 500 errors.

## Tests

Focused tests were rerun after replaying the unchanged patch onto current upstream `main`:

```bash
.venv/bin/python -m pytest \
  tests/models/test_deepseek_v4_amd_moe.py \
  tests/kernels/moe/test_topk_softplus_sqrt.py::test_sqrtsoftplus_bias_uses_deepseek_v4_routing_method \
  tests/kernels/moe/test_topk_softplus_sqrt.py::test_deepseek_v4_fused_shared_expert_is_appended_after_routing \
  tests/kernels/moe/test_topk_softplus_sqrt.py::test_shared_expert_count_preserves_aiter_and_backend_neutral_gates \
  -v
```

Result: **47 passed**.

```bash
.venv/bin/pre-commit run --files \
  tests/kernels/moe/test_topk_softplus_sqrt.py \
  tests/models/test_deepseek_v4_amd_moe.py \
  vllm/_aiter_ops.py \
  vllm/model_executor/layers/fused_moe/experts/rocm_aiter_moe.py \
  vllm/models/deepseek_v4/amd/model.py
```

Result: **all changed-file hooks passed**, including Ruff, formatting, mypy, SPDX, import, configuration, and ROCm API guards. `git diff --check upstream/main...HEAD` also passed.

## Stability scope

During follow-up stress triage, a two-wave high-concurrency GPU memory fault was reproduced on the pre-FSE vLLM parent with official AITER `v0.1.16.post3`, FSE absent, and custom all-reduce disabled (PYNCCL). Wave 1 completed 64/64 requests and wave 2 faulted. That baseline issue is independent of this patch and is tracked in Fangzhou-Ai/vllm#13; it is not presented as an FSE regression or as something this PR fixes.

## AI assistance disclosure

OpenAI Codex assisted with implementation, rebase, test execution, evaluation analysis, and drafting this description. The human submitter reviewed every changed line and is responsible for understanding and defending the change end-to-end.

### CODE REVIEW

File: vllm/models/deepseek_v4/amd/model.py [hunk 1]
```diff
@@ -56,8 +60,281 @@
...
+
+def _should_fuse_shared_expert(vllm_config: VllmConfig) -> bool:
+    config = vllm_config.model_config.hf_config
+    quant_config = vllm_config.quant_config
+    parallel_config = vllm_config.parallel_config
+    reasons = []
+
+    if not current_platform.is_rocm() or not on_gfx950():
+        reasons.append("the device is not ROCm gfx950")
+    if not rocm_aiter_ops.is_fused_moe_enabled():
```
Comment (tjtanaa) [inline vllm/models/deepseek_v4/amd/model.py]: We will also need to make sure that this feature is optional as this is trading off accuracy for speed in general cases. GSM8K is not sufficient to say it is fine in general. Until we have validated in a more comprehensive test suites e.g. NIAH, Agentic workload, it will be best to keep this feature optional. So, we will need to be able to control the feature through `VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS` .

There is a general trend where shared expert quantization could degrade accuracy. We have tried with minimaxm3 fp8 weights, there are slight noticable accuracy degradation. Moreover, the general recipes like on NVIDIA is to not quantize the shared experts due to accuracy reasons.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -75,22 +75,21 @@ def determine_expert_counts(
     num_redundant_experts: int,
     n_shared_experts: int | None,
     is_act_and_mul: bool,
+    enable_shared_expert_fusion: bool | None = None,
```
Comment (tjtanaa) [inline vllm/model_executor/layers/fused_moe/layer.py]: We shouldn't add this argument as we still must rely on

```
      rocm_aiter_ops.is_fusion_moe_shared_experts_enabled()
      or envs.VLLM_ROCM_USE_AITER_FUSION_SHARED_EXPERTS
```

File: vllm/models/deepseek_v4/amd/model.py (hunk 1 above)
Comment (tjtanaa) [reply vllm/models/deepseek_v4/amd/model.py]: `rocm_aiter_ops.is_fusion_moe_shared_experts_enabled`

**Comment by tjtanaa** [comment]
This approach introduces accuracy degradation at a longer input, not even at medium or long context length.


| Configuration | Strict match | Flexible extract | Visible EOF continuations |
| --- | ---: | ---: | ---: |
| FSE disabled, 03:05 UTC | 1253/1319 = **0.9500** | 1252/1319 = **0.9492** | 0 |
| FSE enabled, original conversion, 01:26 UTC | 1242/1319 = **0.9416** | 1137/1319 = **0.8620** | 112 |

**Comment by tjtanaa** [comment]
This is a triage that I could extract out empirically. With the help of Codex The following is the triage report:




# DeepSeek V4 Pro fused-shared-expert accuracy triage

## Conclusion

The controlled A/B result points to fused shared expert (FSE), not to a wrong
token ID.

| Configuration | Strict match | Flexible extract | Visible EOF continuations |
| --- | ---: | ---: | ---: |
| FSE disabled, 03:05 UTC | 1253/1319 = **0.9500** | 1252/1319 = **0.9492** | 0 |
| FSE enabled, original conversion, 01:26 UTC | 1242/1319 = **0.9416** | 1137/1319 = **0.8620** | 112 |
| FSE enabled, AITER-matched conversion, 03:27 UTC | 1242/1319 = **0.9416** | 1139/1319 = **0.8635** | 106 |

The FSE-on and FSE-off runs used the same model, tokenizer, 20-shot prompts,
temperature zero, TP=8, AITER MoE backend, and lm-eval command. The FSE
environment flag is the material model-path difference.

FSE causes two effects:

1. A real answer-quality regression: strict match falls by 11 examples, or
   0.84 percentage points.
2. A much larger benchmark-extraction regression: the changed generation
   trajectory frequently emits `<｜end▁of▁file｜>` and continues into numeric
   dataset text. `strict-match` retains the answer after `####`, while
   `flexible-extract` often selects a later number.

The token is therefore a downstream symptom and score amplifier. It is not the
initiating cause: the official model metadata has always used EOS ID 1, and the
same metadata produces no visible EOF continuations in the FSE-off run.

## FSE mechanism that changes accuracy

FSE is not a numerically equivalent kernel fusion for this checkpoint.

The unfused shared MLP executes the checkpoint's serialized FP8 E4M3 weights
with 128x128 E8M0 block scales. To append that shared expert to the routed
AITER MoE, the FSE loader first dequantizes each shared `w1`, `w2`, and `w3`
matrix and requantizes it to MXFP4 E2M1 with 1x32 scales. The resulting shared
weights have only 4-bit precision.

Measurements on the real layer-3 shared weights show about 12% relative L2
weight error after this conversion:

| Projection | Original conversion | AITER-matched conversion |
| --- | ---: | ---: |
| w1 | 12.240% | 11.962% |
| w3 | 12.221% | 11.941% |
| w2 | 12.321% | 12.032% |

That conversion occurs in every MoE layer, so deterministic temperature-zero
decoding still follows a different token trajectory.

## Conversion mismatch found and tested

The original hand-written converter used `ceil(log2(amax / 6))` for MXFP4
scale selection. AITER's `dynamic_mxfp4_quant` instead rounds the block maximum
to a power of two and divides by four. On actual layer-3 shared weights, these
methods differed in roughly 14-16% of packed values and scales.

The local patch in
`vllm/models/deepseek_v4/amd/model.py` now matches AITER bit-for-bit and reduces
weight reconstruction error. The corresponding zero-block regression test was
updated in `tests/models/test_deepseek_v4_amd_moe.py`.

However, the complete 1,319-sample verification did not recover strict
accuracy: it remained exactly 1242/1319. Flexible extract improved by only two
examples. The scale-selection mismatch was real but was not the principal
score regression.




Codex also found the following observations (AITER-matched conversion):

```diff
diff --git a/tests/models/test_deepseek_v4_amd_moe.py b/tests/models/test_deepseek_v4_amd_moe.py
index e8de3d944..f0bd1fce2 100644
--- a/tests/models/test_deepseek_v4_amd_moe.py
+++ b/tests/models/test_deepseek_v4_amd_moe.py
@@ -224,7 +224,7 @@ def test_deepseek_v4_fp8_to_mxfp4_uses_round_to_nearest_even_bitwise():
     assert torch.equal(raw_packed_scale, packed_scale)
 
 
-def test_deepseek_v4_fp8_to_mxfp4_uses_reference_zero_scale():
+def test_deepseek_v4_fp8_to_mxfp4_matches_aiter_zero_scale():
     weight = torch.zeros((128, 128), dtype=torch.float8_e4m3fn)
     scale = torch.ones(1, 1).to(torch.float8_e8m0fnu)
 
@@ -233,7 +233,8 @@ def test_deepseek_v4_fp8_to_mxfp4_uses_reference_zero_scale():
     )
 
     assert torch.count_nonzero(packed) == 0
-    assert torch.all(packed_scale == 1)
+    # AITER encodes an all-zero block with the minimum E8M0 exponent.
+    assert torch.all(packed_scale == 0)
 
 
 def _make_routed_experts(
diff --git a/vllm/models/deepseek_v4/amd/model.py b/vllm/models/deepseek_v4/amd/model.py
index 969d4a0d1..4b23435da 100644
--- a/vllm/models/deepseek_v4/amd/model.py
+++ b/vllm/models/deepseek_v4/amd/model.py
@@ -168,8 +168,15 @@ def _fp8_block_weight_to_mxfp4(
     mxfp4_block_size = 32
     blocks = weight_bf16.view(-1, mxfp4_block_size)
     block_amax = blocks.float().abs().amax(dim=-1, keepdim=True)
-    block_amax.clamp_(min=6.0 * (2**-126))
-    scale_exp = (block_amax * (1.0 / 6.0)).log2().ceil()
+    # Match AITER's dynamic_mxfp4_quant scale selection.  MXFP4's E8M0
+    # scale is chosen by rounding amax to a power of two and dividing by
+    # four.  Using ceil(log2(amax / 6)) is a valid no-saturation encoding,
+    # but it disagrees with the quantizer used by the AITER MoE path and
+    # produces measurably more error for the DeepSeek V4 shared weights.
+    amax_bits = block_amax.view(torch.int32)
+    rounded_amax_bits = (amax_bits + 0x200000) & 0xFF800000
+    rounded_amax = rounded_amax_bits.view(torch.float32)
+    scale_exp = rounded_amax.log2().floor().sub_(2.0)
     scale_exp.clamp_(-127.0, 127.0)
     blocks.div_(torch.exp2(scale_exp))
```

**Comment by tjtanaa** [comment]
If there are no proper solution, then we will need to add a warning.

**Comment by Fangzhou-Ai** [comment]
@tjtanaa  a native heterogeneous kernel is added to aiter to fully mitigate the precision issue while preserves the perf. We need to bump the aiter first then re-visit this PR, labeling it as draft for now.

**Comment by tjtanaa** [comment]
Thank you @Fangzhou-Ai . I will prioritize is merging this PR first then https://github.com/vllm-project/vllm/pull/48044 .

**Comment by Fangzhou-Ai** [comment]
E2E GSM8K 8*MI355X DSV4 Pro TP=8 EP=1 test result, conc=64 n-shot=20
image
